requests/models.py

```diff
@@ -455,28 +443,41 @@ def prepare_body(self, data, files, json=None):
                     else:
                         content_type = 'application/x-www-form-urlencoded'
 
-            self.prepare_content_length(body)
-
             # Add content-type if it wasn't explicitly provided.
             if content_type and ('content-type' not in self.headers):
                 self.headers['Content-Type'] = content_type
 
+        self.prepare_content_length(body)
         self.body = body
 
     def prepare_content_length(self, body):
-        if hasattr(body, 'seek') and hasattr(body, 'tell'):
-            curr_pos = body.tell()
-            body.seek(0, 2)
```

Comment:
My impression at the time was that it would produce the same calculation results, but the calculation is done slightly differently. @kennethreitz you raise a good point - using `len(o.getvalue())` in `super_len` seems like it could be less efficient than using seek/tell to calculate the length. I'm not sure if it would make sense to change this in `super_len`? Seems like that should be a separate PR though.